File: scripts/julia_controller/fleet_server.jl
# fleet_server.jl - Fleet mode: bridges RIFT's HTTP protocol
# (https://github.com/CursedPrograms/RIFT) to fleet_state.angles, which the
# main loop reads each frame and sends over serial like any other mode. Also
# serves the scripts/web/ control page, so the arm can be driven from any
# browser on the network, not just RIFT. Julia counterpart of controller.py's
# create_fleet_app()/start_fleet_heartbeat() and the C++ controller's
# fleet_server.cpp - same endpoints, same wire protocol.

import HTTP
import JSON3
import Sockets

const FLEET_NAME = "ARM"
const FLEET_TYPE = "robot"
const FLEET_CAPABILITIES = ["servo_control", "6dof", "arm"]
const DEFAULT_FLEET_PORT = 5011
# Must stay under RIFT's FLEET_TTL_SECS (20s).
const FLEET_HEARTBEAT_SECS = 10

mutable struct FleetState
    angles::Dict{Int,Int}
    lock::ReentrantLock
end
FleetState(angles::Dict{Int,Int}) = FleetState(angles, ReentrantLock())

function local_lan_ip()
    try
        return string(Sockets.getipaddr())
    catch
        return "127.0.0.1"
    end
end

function content_type_for(path)
    endswith(path, ".html") && return "text/html; charset=utf-8"
    endswith(path, ".css") && return "text/css; charset=utf-8"
    endswith(path, ".js") && return "application/javascript; charset=utf-8"
    return "application/octet-stream"
end

function serve_file(full_path::AbstractString, url_path::AbstractString)
    isfile(full_path) || return HTTP.Response(404, "Not found")
    body = read(full_path)
    return HTTP.Response(200, ["Content-Type" => content_type_for(url_path)], body)
end

json_response(status::Int, obj) =
    HTTP.Response(status, ["Content-Type" => "application/json"], JSON3.write(obj))

function make_handler(motors, fleet_state::FleetState, connected_fn::Function, web_dir::AbstractString)
    return function (req::HTTP.Request)
        uri = HTTP.URI(req.target)
        path = uri.path

        if path == "/"
            return serve_file(joinpath(web_dir, "index.html"), "/index.html")
        elseif path == "/style.css" || path == "/app.js"
            return serve_file(joinpath(web_dir, path[2:end]), path)
        elseif path == "/ping"
            return HTTP.Response(200, "$FLEET_NAME alive")
        elseif path == "/status"
            motors_json = Dict{String,Any}()
            lock(fleet_state.lock) do
                for (n, m) in motors
                    motors_json[string(n)] = Dict(
                        "channel" => m["channel"], "angle" => fleet_state.angles[n],
                        "min" => m["min"], "max" => m["max"],
                    )
                end
            end
            return json_response(200, Dict("connected" => connected_fn(), "motors" => motors_json))
        elseif path == "/cmd"
            q = HTTP.queryparams(uri)
            motor = tryparse(Int, get(q, "motor", ""))
            angle = tryparse(Int, get(q, "angle", ""))
            if motor === nothing || angle === nothing
                return json_response(400, Dict("error" => "expected ?motor=<1-6>&angle=<degrees>"))
            elseif !haskey(motors, motor)
                return json_response(400, Dict("error" => "unknown motor $motor"))
            else
                m = motors[motor]
                angle = clamp(angle, m["min"], m["max"])
                lock(fleet_state.lock) do
                    fleet_state.angles[motor] = angle
                end
                return json_response(200, Dict("motor" => motor, "angle" => angle))
            end
        elseif path == "/reset"
            q = HTTP.queryparams(uri)
            if haskey(q, "motor")
                motor = tryparse(Int, q["motor"])
                if motor === nothing || !haskey(motors, motor)
                    return json_response(400, Dict("error" => "unknown motor $(get(q, "motor", ""))"))
                end
                rest_angle = motors[motor]["rest"]
                lock(fleet_state.lock) do
                    fleet_state.angles[motor] = rest_angle
                end
                return json_response(200, Dict("motor" => motor, "angle" => rest_angle))
            else
                resp = Dict{String,Any}()
                lock(fleet_state.lock) do
                    for (n, m) in motors
                        fleet_state.angles[n] = m["rest"]
                        resp[string(n)] = m["rest"]
                    end
                end
                return json_response(200, resp)
            end
        else
            return HTTP.Response(404, "Not found")
        end
    end
end

function start_fleet_heartbeat(rift_host::AbstractString, rift_port::Int; interval::Real=FLEET_HEARTBEAT_SECS)
    @async begin
        while true
            try
                HTTP.post(
                    "http://$rift_host:$rift_port/register",
                    ["Content-Type" => "application/x-www-form-urlencoded"],
                    HTTP.escapeuri(Dict(
                        "name" => FLEET_NAME, "type" => FLEET_TYPE,
                        "capabilities" => join(FLEET_CAPABILITIES, ","),
                    ));
                    readtimeout=2, connect_timeout=2,
                )
            catch
                # RIFT/NORA not reachable yet - keep retrying.
            end
            sleep(interval)
        end
    end
end

"""
    start_fleet_server(motors, fleet_state, connected_fn; port, register, rift_host, rift_port, web_dir)

Starts the Fleet-mode HTTP bridge (non-blocking) and, if `register`, a
background RIFT/NORA heartbeat. Returns "" on success, or an error message
(e.g. port already in use) - never throws, so Fleet mode failing doesn't
crash the rest of the controller.
"""
function start_fleet_server(motors, fleet_state::FleetState, connected_fn::Function;
                             port::Int=DEFAULT_FLEET_PORT, register::Bool=true,
                             rift_host::AbstractString="127.0.0.1", rift_port::Int=5000,
                             web_dir::AbstractString)
    try
        probe = Sockets.listen(Sockets.InetAddr(Sockets.ip"0.0.0.0", port))
        close(probe)
    catch e
        return "Could not bind to port $port (already in use?): $e"
    end

    handler = make_handler(motors, fleet_state, connected_fn, web_dir)
    try
        # HTTP.serve! is itself non-blocking (spawns its own background
        # accept loop and returns immediately) - calling it directly here
        # means a synchronous startup failure is actually caught below.
        HTTP.serve!(handler, "0.0.0.0", port; verbose=false)
    catch e
        return "Failed to start Fleet HTTP server: $e"
    end

    register && start_fleet_heartbeat(rift_host, rift_port)
    return ""
end

File: scripts/csharp_controller/FleetServer.cs
// FleetServer.cs - Fleet mode: bridges RIFT's HTTP protocol
// (https://github.com/CursedPrograms/RIFT) to FleetState.Angles, which the UI
// tick reads each frame and sends over serial like any other mode. Also serves
// the scripts/web/ control page so the arm can be driven from any browser on
// the network. Same endpoints and wire protocol as controller.py's Flask app
// and the C++ controller's fleet_server.cpp.

using System.Net;
using System.Net.Sockets;
using Microsoft.AspNetCore.Builder;
using Microsoft.AspNetCore.Hosting;
using Microsoft.AspNetCore.Http;
using Microsoft.Extensions.Hosting;
using Microsoft.Extensions.Logging;

namespace ArmController;

public sealed class FleetState
{
    public readonly object Lock = new();
    public readonly Dictionary<int, int> Angles = new();
}

public sealed class FleetServer : IDisposable
{
    const string FleetName = "ARM";
    const string FleetType = "robot";
    const string FleetCapabilities = "servo_control,6dof,arm";
    public const int DefaultPort = 5011;
    static readonly TimeSpan HeartbeatInterval = TimeSpan.FromSeconds(10); // must stay under RIFT's 20s TTL

    WebApplication? _app;
    CancellationTokenSource? _cts;

    /// <summary>Best-effort LAN IP, so the UI can show an address RIFT (on another device) can reach.</summary>
    public static string LocalLanIp()
    {
        try
        {
            using var s = new Socket(AddressFamily.InterNetwork, SocketType.Dgram, ProtocolType.Udp);
            s.Connect("8.8.8.8", 80); // UDP connect: no traffic is actually sent
            return ((IPEndPoint)s.LocalEndPoint!).Address.ToString();
        }
        catch
        {
            return "127.0.0.1";
        }
    }

    /// <summary>Starts the HTTP bridge (and RIFT heartbeat if requested). Returns "" on success, else an error message.</summary>
    public string Start(IReadOnlyDictionary<int, Motor> motors, FleetState state, Func<bool> connected,
                        int port, bool register, string riftHost, int riftPort)
    {
        try
        {
            var builder = WebApplication.CreateBuilder();
            builder.Logging.ClearProviders();
            builder.WebHost.UseUrls($"http://0.0.0.0:{port}");
            var app = builder.Build();
            MapRoutes(app, motors, state, connected);
            Task.Run(() => app.StartAsync()).GetAwaiter().GetResult();
            _app = app;
        }
        catch (Exception e)
        {
            return $"Could not start Fleet server on port {port} (already in use?): {e.Message}";
        }

        if (register)
        {
            _cts = new CancellationTokenSource();
            _ = Task.Run(() => HeartbeatLoop(riftHost, riftPort, _cts.Token));
        }
        return "";
    }

    static IResult ServeFile(string name, string contentType)
    {
        string path = Path.Combine(Config.WebDir, name);
        return File.Exists(path) ? Results.Bytes(File.ReadAllBytes(path), contentType) : Results.NotFound("Not found");
    }

    static IResult Error(string message) => Results.Json(new Dictionary<string, object> { ["error"] = message }, statusCode: 400);

    static void MapRoutes(WebApplication app, IReadOnlyDictionary<int, Motor> motors, FleetState state, Func<bool> connected)
    {
        app.MapGet("/", () => ServeFile("index.html", "text/html; charset=utf-8"));
        app.MapGet("/style.css", () => ServeFile("style.css", "text/css; charset=utf-8"));
        app.MapGet("/app.js", () => ServeFile("app.js", "application/javascript; charset=utf-8"));
        app.MapGet("/ping", () => Results.Text($"{FleetName} alive", "text/plain"));

        app.MapGet("/status", () =>
        {
            var motorsJson = new Dictionary<string, object>();
            lock (state.Lock)
            {
                foreach (var (n, m) in motors)
                    motorsJson[n.ToString()] = new Dictionary<string, object>
                    {
                        ["channel"] = m.Channel, ["angle"] = state.Angles[n], ["min"] = m.Min, ["max"] = m.Max,
                    };
            }
            return Results.Json(new Dictionary<string, object> { ["connected"] = connected(), ["motors"] = motorsJson });
        });

        app.MapGet("/cmd", (HttpRequest req) =>
        {
            if (!int.TryParse(req.Query["motor"].ToString(), out int motor) || !int.TryParse(req.Query["angle"].ToString(), out int angle))
                return Error("expected ?motor=<1-6>&angle=<degrees>");
            if (!motors.TryGetValue(motor, out var m))
                return Error($"unknown motor {motor}");
            angle = Math.Clamp(angle, m.Min, m.Max);
            lock (state.Lock) state.Angles[motor] = angle;
            return Results.Json(new Dictionary<string, object> { ["motor"] = motor, ["angle"] = angle });
        });

        app.MapGet("/reset", (HttpRequest req) =>
        {
            if (req.Query.ContainsKey("motor"))
            {
                if (!int.TryParse(req.Query["motor"].ToString(), out int motor) || !motors.TryGetValue(motor, out var m))
                    return Error($"unknown motor {req.Query["motor"]}");
                lock (state.Lock) state.Angles[motor] = m.Rest;
                return Results.Json(new Dictionary<string, object> { ["motor"] = motor, ["angle"] = m.Rest });
            }

            var all = new Dictionary<string, object>();
            lock (state.Lock)
            {
                foreach (var (n, m) in motors)
                {
                    state.Angles[n] = m.Rest;
                    all[n.ToString()] = m.Rest;
                }
            }
            return Results.Json(all);
        });
    }

    static async Task HeartbeatLoop(string host, int port, CancellationToken ct)
    {
        using var http = new HttpClient { Timeout = TimeSpan.FromSeconds(2) };
        string url = $"http://{host}:{port}/register";
        while (!ct.IsCancellationRequested)
        {
            try
            {
                using var form = new FormUrlEncodedContent(new Dictionary<string, string>
                {
                    ["name"] = FleetName, ["type"] = FleetType, ["capabilities"] = FleetCapabilities,
                });
                using var _ = await http.PostAsync(url, form, ct);
            }
            catch
            {
                // RIFT/NORA not reachable yet - keep retrying.
            }

            try { await Task.Delay(HeartbeatInterval, ct); }
            catch (OperationCanceledException) { break; }
        }
    }

    public void Dispose()
    {
        _cts?.Cancel();
        if (_app != null)
        {
            try { Task.Run(() => _app.StopAsync(TimeSpan.FromSeconds(1))).Wait(TimeSpan.FromSeconds(2)); }
            catch { /* shutting down anyway */ }
        }
    }
}
